### ISSUE TITLE
[HOLD #12054] [$1000] Conversation DM bounces, flickers, and lags when scrolling downward in a long chat

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#2545]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
Scrolling through a chat DM should be smooth (doesn't bounce, flicker, or lag) regardless of conversation length.

## Actual Result:
The chat DM bounces when scrolling downward in a long conversation.

## Action Performed:
1. Open the Expensify.cash app on an Android device
2. Navigate to a chat DM that has a long conversation of back and forth.
3. Scroll upwards for a few seconds.
4. Stop scrolling upward, and then slowly scroll downard at a reading speed. 

Note: lag and flickering behavior is especially worsened if you:
- Scroll slowly by tapping and releasing your finger each time (i.e. after each scroll downward, lift your finger, wait a moment, and then repeat...)
- Fully expand the keyboard and scroll downward slowly.

## Workaround:
You can still scroll, but it's a poor UX and the scroll should be smooth.

## Platform:
>Where is this issue occurring?
All platforms

 Web
 iOS
 Android 
 Desktop App
 Mobile Web

**Version Number:** 1.0.65-0

**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/44479856/115792143-beb73900-a397-11eb-9183-c9f9dc35da61.mp4

**Expensify/Expensify Issue URL:** https://www.upwork.com/jobs/~0140ca576a3b244cff

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

___ 
From @quinthar https://expensify.slack.com/archives/C01GTK53T8Q/p1619067337008500
>ISSUE: scrolling backwards seems to work great, but when you scroll forwards it Jitters all over the place. It's very hard to describe, but it's perfectly reproducible on android. Just go into a long conversation, scroll back a few pages, and then scroll forward slowly, and it bounces all over the place.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mateocole (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @ctkochan22 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by ctkochan22** [comment]
Labels look good

**Comment by mallenexpensify** [comment]
@ctkochan22 bouncing back to you - do you think this can be worked on by a Contributor?

**Comment by ctkochan22** [comment]
@mallenexpensify Shoot sorry. Yes I think so.

**Comment by mallenexpensify** [comment]
Thanks @ctkochan22 , in the future, if a contributor can work on an issue apply the `External` issue so it can be posted to Upwork. Adding it now.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @michaelhaxhiu (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by michaelhaxhiu** [comment]
Will post today.

**Comment by michaelhaxhiu** [comment]
Just tested on my Android, and I was able to reproduce successfully. Posted to Upwork as of right now ✅ .

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Julesssss (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by Julesssss** [comment]
CC @kidroca, as he had some ideas for[ these improvements here](https://github.com/Expensify/react-native-onyx/pull/76#issuecomment-853789066). Perhaps you would be interested in this issue?

**Comment by kidroca** [comment]
I don't think this is related to cache. 

When I studied the effects of virtualization I came to a similar situation when I had a large chat: https://github.com/Expensify/Expensify.cash/issues/2629#issuecomment-832516464

My theory (short version) 
For big chats you reach the virtualization window boundary
As you scroll down an item is removed from the top
Another item is added on the bottom
Native, does not provide `getItemLayout` to the flat list
It has to put the item, render it and then shift content if it has to (only when scrolling down)

Implementing a fast `getItemLayout` would probably be hard. 
There's an open ticket about a replacement of the FlatList component which might solve the problem here entirely

In theory this should be recreatable on iOS as well, maybe the list handles better there as the devices are beefier. 
Any chance someone could try this on the oldest iPhone supported?

**Comment by michaelhaxhiu** [comment]
>Any chance someone could try this on the oldest iPhone supported?

This would be good to ask in #expensify-open-source I think, as we have more eyes and exposure there. Can you post it there @kidroca? I agree with the idea of seeing if old iOS devices are included in the scope of the issue.

**Comment by rdjuric** [comment]
@kidroca @michaelhaxhiu 
Here's a test running on an old iPhone 5s.

https://user-images.githubusercontent.com/53711423/122099933-f6b38900-cde8-11eb-85a5-ec7a9b97f9ec.mov

It's from the build in the app store, so probably a few versions behind our main. Feels smooth on iOS and doesn't flicker.

**Comment by michaelhaxhiu** [comment]
@rdjuric thank you! Can you try once more, but this time expand our keyboard and scroll downward slowly? If it behaves the same, no need to record another vid.

**Comment by rdjuric** [comment]
@michaelhaxhiu Sorry! Forgot about the expanded keyboard 😅


https://user-images.githubusercontent.com/53711423/122101860-2d8a9e80-cdeb-11eb-872f-c0458940cc47.mp4

The issue does happens on iOS. I was also able to reproduce it while scrolling up the chat, not only down, seems to happen whenever the FlatList render new items. It stops flickering when we approach the end (or beginning) of the chat, as the FlatList doesn't need to render any new items.

**Comment by michaelhaxhiu** [comment]
Any takers on this issue so far?

**Comment by kidroca** [comment]
IMO this should wait the result of #2985 
I don't see an easy fix and this might result in double or unnecessary work if the chat scrolling implementation is to be rehauled
Proper Bidirectional scrolling implementation should fix the current issue

**Comment by michaelhaxhiu** [comment]
Applying the `planning` label till https://github.com/Expensify/Expensify.cash/issues/2985 is released.

**Comment by michaelhaxhiu** [comment]
Still waiting on https://github.com/Expensify/Expensify.cash/issues/2985

**Comment by michaelhaxhiu** [comment]
Still waiting on https://github.com/Expensify/Expensify.cash/issues/2985

**Comment by michaelhaxhiu** [comment]
Still waiting on https://github.com/Expensify/Expensify.cash/issues/2985

**Comment by MelvinBot** [comment]
@Julesssss, @michaelhaxhiu it looks like no one is assigned to work on this job.
Please double the price or add a comment stating why the job isn't being doubled.

**Comment by michaelhaxhiu** [comment]
Still blocked. Can I just close this GH since we are already tracking this in E/App here?
https://github.com/Expensify/Expensify/issues/167383

**Comment by michaelhaxhiu** [comment]
Still on hold.

**Comment by michaelhaxhiu** [comment]
Rory started working on https://github.com/Expensify/App/issues/2985 and will post an update sometime early this week.

**Comment by michaelhaxhiu** [comment]
Awaiting an update from Rory on https://github.com/Expensify/App/issues/2985, which Jeremy requested yesterday (September 8 2021).

**Comment by michaelhaxhiu** [comment]
Rory is making good progress on bi-directional scrolling, and working with Kidroca on https://github.com/Expensify/App/issues/2985

**Comment by Julesssss** [comment]
Issue is blocked currently

**Comment by Julesssss** [comment]
No change

**Comment by Julesssss** [comment]
Issue is still on hold due as we're awaiting the outcome of this issue: #2985

**Comment by Julesssss** [comment]
It looks like a solution is not far away in the parent issue.

**Comment by michaelhaxhiu** [comment]
Still keeping this one in the ol' holster until parent GH is complete.

**Comment by Julesssss** [comment]
Still waiting on the above

**Comment by michaelhaxhiu** [comment]
Luca & Rory are making gradual progress on the blocker GH. 🤞

**Comment by michaelhaxhiu** [comment]
Still blocked.

**Comment by michaelhaxhiu** [comment]
Blocked still

**Comment by michaelhaxhiu** [comment]
Hold

**Comment by Julesssss** [comment]
Still held

**Comment by michaelhaxhiu** [comment]
still holding on https://github.com/Expensify/App/issues/2985

**Comment by michaelhaxhiu** [comment]
Not overdue

**Comment by michaelhaxhiu** [comment]
Still on hold

**Comment by michaelhaxhiu** [comment]
Still holding

**Comment by michaelhaxhiu** [comment]
still holding

**Comment by michaelhaxhiu** [comment]
This is holding on birectional scrolling still.

**Comment by trjExpensify** [comment]
Didn't we complete the bidirectional scrolling issue, so can't this come off hold now or are we actually held on comment linking or something?

**Comment by JmillsExpensify** [comment]
Yeah agreed, I don't think this issue should on hold anymore.

**Comment by michaelhaxhiu** [comment]
Ugh damn, definitely got confused by the titling of https://github.com/Expensify/App/issues/2985 (which is actually about comment linking).

**Comment by michaelhaxhiu** [comment]
Going to take the liberty to price this at $500 to start. cc @kidroca in case you want to close a cold case 🕵️ 💼

**Comment by michaelhaxhiu** [comment]
@Expensify/applause Can someone please test this issue again to ensure it's still reproducible on **Android**?

**Comment by JmillsExpensify** [comment]
Thanks for getting this one going again!

**Comment by kidroca** [comment]
I was able to recreate this on Android Emulator
Checkout around the 4th and the 20th second marks:

https://user-images.githubusercontent.com/12156624/194669183-750c04a3-a022-443e-8997-eed8900ad361.mp4

I'm not sure what causes this, but it's probably related to virtualization and items being rendered outside the view window 
As we scroll down actions are removed from the top and added to the bottom which causes a flicker, but it's very hard to recreate and trace

**Comment by michaelhaxhiu** [comment]
>but it's probably related to virtualization 

Does that invalidate the reproduction, or suggest that the problem still exists? I just reproduced it on iOS device (latest version) in a DM with a colleague. I can't upload the video because it has some sensitive internal information... but working on reproducing again in a test env where I can share the video. But safe to say - this definitely still exists and needs to be fixed.

**Comment by JmillsExpensify** [comment]
I agree this happens pretty frequently, and in fact, I can produce it in either scroll direction pretty easily on a chat that has lots of history. Here's an example with Concierge. 

https://user-images.githubusercontent.com/1127863/194794715-671f49d1-62bd-4552-89bb-ff50c7f09f96.mov

**Comment by michaelhaxhiu** [comment]
I'm not sure if this is resolvable **Externally** at this point in time, but I think it's valuable to figure it out and pay any parties who help us confirm. cc @Julesssss

**Comment by Julesssss** [comment]
@michaelhaxhiu we have quite a few reproductions, let's leave the issue open and see if we receive any proposals.

**Comment by kidroca** [comment]
@michaelhaxhiu 
> > but it's probably related to virtualization
> 
> Does that invalidate the reproduction video in any way?

No, I'm just thinking of what might be causing the problem 
We can test whether it's virtualization related, by disabling virtualization for the inverted list, though this might have a big performance hit and make it impossible to scroll anyway 

Since it happens on iOS as well I lean towards virtualization and/or how we render report action items - rendering an item adjusts the virtual list on the fly (item scroll position, list scrollable height), because the list has no idea how big is a comment before it render it


@JmillsExpensify 
> I agree this happens pretty frequently, and in fact, I can produce it in either scroll direction pretty easily on a chat that has lots of history.

Your case might be related to this ticket
- https://github.com/Expensify/App/issues/7860

**Comment by JmillsExpensify** [comment]
@kidroca Oh thanks, that one wasn't on my radar. Hopefully we'll start to get movement on it soon since we've started work on comment linking, which blocks it in turn.

**Comment by michaelhaxhiu** [comment]
Doubling price to $1000, and re-surfacing this comment:

>I'm not sure if this is resolvable Externally at this point in time, but I think it's valuable to figure it out and pay any parties who help us confirm.

**Comment by JmillsExpensify** [comment]
Given that we seem to be unholding #7860, should we now put this issue on [Hold] since it will likely be solved by that issue? That's an epic proposal that is going to be broken down into 4-5 separate issues.

**Comment by michaelhaxhiu** [comment]
@JmillsExpensify my only hesitation is assuming this will certainly be fixed by #7860. 

Or perhaps better - we can directly ask @LucioChavezFuentes if they agree that this problem will be fixed via their proposal in https://github.com/Expensify/App/issues/7860.

**Comment by michaelhaxhiu** [comment]
Adding Hold to GH title as we are tracking this issue as part of https://github.com/Expensify/App/issues/12054

**Comment by LucioChavezFuentes** [comment]
> Or perhaps better - we can directly ask @LucioChavezFuentes if they agree that this problem will be fixed via their proposal in https://github.com/Expensify/App/issues/7860.

On Web I'm sure this problem will be fixed with my proposal. But in Android and IOS can't reproduce the problem. Need to make more tests on these platforms. Do the lists with issues have images? 

By the way, what 'DM' means?

**Comment by michaelhaxhiu** [comment]
@LucioChavezFuentes ah sorry, DM = direct message

**Comment by michaelhaxhiu** [comment]
Still held.

**Comment by michaelhaxhiu** [comment]
Same same melvin

**Comment by michaelhaxhiu** [comment]
Held still. Keeping assigned while I'm OOO for 1.5 weeks. If this becomes unblocked please re-assign a new BZ @Julesssss

**Comment by Julesssss** [comment]
Still held

**Comment by Julesssss** [comment]
No change

**Comment by Santhosh-Sellavel** [comment]
on hold

**Comment by michaelhaxhiu** [comment]
Stillllll held

**Comment by Julesssss** [comment]
Held stilll

**Comment by Santhosh-Sellavel** [comment]
onhold

**Comment by JmillsExpensify** [comment]
What's our plan for #12054? Do either of you know?

**Comment by Julesssss** [comment]
bumped the linked issue for an update

**Comment by Julesssss** [comment]
We seem held up on the contributor.

**Comment by michaelhaxhiu** [comment]
Posting internally to see if we ought to revise our gameplan or hold course.

https://expensify.slack.com/archives/C01SKUP7QR0/p1677692378964509

**Comment by Santhosh-Sellavel** [comment]
On hold

**Comment by michaelhaxhiu** [comment]
Still in progress... 

Janic from Margelo mentioned:
>I had a look, the solution here https://github.com/Expensify/react-native-web/commit/ab2ebd84250d60598e69f24c7289aafd1e82f659 seems good and what the maintainer of RN web is looking for.
>The main idea is to fix RN web behaviours to match RN exactly to fix the issues with virtualized list instead of patching virtualized list to work with rn web behaviours.

**Comment by JmillsExpensify** [comment]
Further updates in the linked hold issue this week.

**Comment by Julesssss** [comment]
[Seems like](https://github.com/Expensify/App/issues/12054#issuecomment-1495372407) the contributor has merged 3 PRs related to pagination! I think we are now waiting for another PR and then for the upstream library to be updated.

**Comment by michaelhaxhiu** [comment]
Progress is 📈 . Waiting for the fix to merge upstream at this point per https://github.com/Expensify/App/issues/12054#issuecomment-1505811916

**Comment by michaelhaxhiu** [comment]
Seems stuff is still progression, and we are very close! Let's see next week if this is ready 🐣

**Comment by michaelhaxhiu** [comment]
@Julesssss and @Santhosh-Sellavel just to confirm, is this still held? I'm having a hard time following if [this is a blocker](https://github.com/Expensify/App/issues/12054#issuecomment-1532187538) or not?

**Comment by Julesssss** [comment]
Yeah we're now holding on the [RN library bump](https://github.com/Expensify/App/issues/12054#issuecomment-1532187538).

**Comment by michaelhaxhiu** [comment]
No movement, still held.

**Comment by michaelhaxhiu** [comment]
Same same.

**Comment by michaelhaxhiu** [comment]
Held

**Comment by michaelhaxhiu** [comment]
Still on hold

**Comment by michaelhaxhiu** [comment]
same same

**Comment by michaelhaxhiu** [comment]
Note: I'm preparing to go OOO for ~2 weeks, and making this `Monthly` because I doubt it'll get finished while I'm gone. When I return I'll flip it back to `Weekly` and oversee it.

**Comment by michaelhaxhiu** [comment]
Back from OOO, switching back to `Weekly` to monitor this and not let it collect dust.

**Comment by Julesssss** [comment]
BLOCKED

**Comment by michaelhaxhiu** [comment]
Still blocked

**Comment by michaelhaxhiu** [comment]
Still holding on pagination https://github.com/Expensify/App/issues/12054

**Comment by michaelhaxhiu** [comment]
I'm re-assigning this as part of my preparation for Sabbatical (starting tomorrow). **Next steps:**

- This GH is stagnant and on hold till Pagination is in motion https://github.com/Expensify/App/issues/12054. 
- Labelling `monthly` for that reason

**Comment by greg-schroeder** [comment]
Still on hold

**Comment by greg-schroeder** [comment]
This has been open for 2.5 years, does this need to be open? 😅

**Comment by Julesssss** [comment]
Lol, probably not? We can always reopen if someone raises this again

**Comment by greg-schroeder** [comment]
Haha okay
